File: dflow/src/app/components/dashboard/services/post-panel.ts
'use client'

import axios from "axios"
import { PanelWraper } from "../dashboard";
import { PanelContentType } from "../panel/panel";
import { UserQuery } from "../../user-query/model/user-query";
import { ChartType } from "../../visualizations/types";

export interface PanelDto {
    model_id: string,
    panel_id: string,
    user_id: string,
    panel: {
        layout: {
            x: number,
            y: number,
            w: number,
            h: number,
            minH: number,
            minW: number,
        },
        props: {
            name: string,
            content_type: {
                Chart: {
                    chart_type: ChartType
                }
            }
        },
        query: UserQuery["query"];
    }
}

export const postPanel = async (panel: PanelWraper, userId: string = "default") => {
    const body: PanelDto = {
        model_id: panel.props.builder.getModel().id,
        panel_id: panel.props.id,
        user_id: userId,
        panel: {
            layout: {
                x: panel.layout.x,
                y: panel.layout.y,
                w: panel.layout.w,
                h: panel.layout.h,
                minH: panel.layout.minH || 0,
                minW: panel.layout.minW || 0,
            },
            props: {
                name: panel.props.name,
                content_type: {
                    Chart: {
                        chart_type: panel.props.content.chartType || "table"
                    }
                }
            },
            query: panel.props.builder.build().query
        }
    }
    return axios.post('http://127.0.0.1:8000/dashboard/panel/save',
        { ...body }
    );
} 

File: dflow_api/src/modules/dashboard/infrastructure/routes/post_panel.rs
use crate::{
    modules::dashboard::{
        infrastructure::persistence::panel_saver::PanelSaver, model::panel::PanelDto,
    },
    template_dir::{http500, Error500Template},
    Db,
};
use rocket::{http::Status, post, serde::json::Json};
use rocket_db_pools::Connection;

///
/// Example query
/// ```text
/// {
///    "model_id": "test",
///    "panel_id": "abcdefgh",
///    "user_id": "test",
///    "panel": {
///      "layout": {
///        "x": 0,
///        "y": 0,
///        "w": 0,
///        "h": 0,
///        "minH": 0,
///        "minW": 0
///      },
///      "props": {
///        "name": "foo",
///        "content_type": {
///             "Chart": {
///                 "chart_type": "bar"
///             }
///        }
///      },
///        "query": null
///    }
///  }
///```
#[post("/panel/save", data = "<panel_dto>")]
pub(crate) async fn post_panel_handler(
    mut db: Connection<Db>,
    panel_dto: Json<PanelDto>,
) -> Result<(), (Status, Error500Template)> {
    let mut panel_saver = PanelSaver::new(&mut db);
    panel_saver
        .insert_panel(&panel_dto)
        .await
        .map_err(|e| http500(e))
}
